# ECBM-4060-Project

## Basic info
This is the final project for ECBM 4060. 
Authors: Mingyuan Zhang, Siddartha Marella. 
Description: An interactive tool for SARS-CoV-2 epidemic trends analysis. 

## Prerequisite and Dependency

Java 10 runtime envirionment is required for proper functioning
Tested on Python 3.8 with pandas, numpy, and matplotlb. May work with higher version. 
The program needs to be executed on Linux platform. 

Bowtie2, BCFTools, VCFTools, SAMTools are required. Make sure they are installed and added to $PATH

## Usage
To use the program, simply lauch Python3 main.py
The program have two modes of operation: "all" and "haplotype"
"all" mode executes a complete analysis from sequences to haplotypes
"haplotype" mode executes a haplotype identification to identify haplotypes from a list of mutation information.
**Note, the debug mdoe is not intended for the user, only use it after backing up your input data.**
For more details, type help after lauching the program


## Inputs
When doing complete analysis, the inputs should be a fasta file or multiple fasta files containing genome sequences of COVID-19. 

When identifying haplotypes from known mutations info, the input should be a single .tsv file containing the mutation ID-POS-REF-ALT infor. Such file can be generated by using this tool in "all" mode. 

## Ouputs
The results will by default be generated in the /output folder, which may be manually change to any writatble directory.
